### PR TITLE
[5.4] Refactor Procedures&Functions + CI Updates

### PR DESCRIPTION
**Pull request by mstaack**

### CODE REVIEW

**Comment by mstaack** [comment]
@yajra i cleaned up the procedures stuff and wrote some tests...

currently you just need to run this docker instance for the functional tests
`docker run -d -p 49160:22 -p 49161:1521 wnameless/oracle-xe-11g` then have a local php / oci8 setup

need to fix formatting issues ... do we have a working codestyle config for csfixer?

what do you think of running this on travis

**Comment by mstaack** [comment]
@yajra i left `createStatementFromProcedure` and `createSqlFromProcedure` public to allowe users in caller side to craft their own bindings etc...  oracle procedure bindings are ugly ;)

let me know what you want edited ...so we can merge this

**Comment by yajra** [comment]
Wow, nice work @mstaack! ❤️ the oci8 build. I think we can do actual DB integration test on this build. Will add this to my sched for review/testing tom. Thanks a lot!

File: src/Oci8/Oci8Connection.php [hunk 1]
```diff
@@ -216,74 +216,143 @@ protected function getDoctrineDriver()
      * 'bye'], PDO::PARAM_LOB).
      *
      * @author Tylerian - jairo.eog@outlook.com
-     * @param string $sql (mixed)
+     *
+     * @param       $functionName
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: `string $functionName`?

File: src/Oci8/Oci8Connection.php [hunk 2]
```diff
@@ -216,74 +216,143 @@ protected function getDoctrineDriver()
...
      * @param array $bindings (kvp array)
-     * @param int $returnType (PDO::PARAM_*)
-     * @param int $length
+     * @param int   $returnType (PDO::PARAM_*)
+     * @param int   $length
+     *
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: Please remove new line before return. That's how laravel does this.

File: src/Oci8/Oci8Connection.php [hunk 3]
```diff
@@ -216,74 +216,143 @@ protected function getDoctrineDriver()
...
 
-            return $results;
-        }
+        return $results;
+    }
+
+    /**
+     * Creates sql command to run a procedure with bindings
+     *
+     * @param             $procedureName
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: Missing type `string`?

File: src/Oci8/Oci8Connection.php [hunk 4]
```diff
@@ -216,74 +216,143 @@ protected function getDoctrineDriver()
...
+        $prefix = count($bindings) ? ',' : '';
+        $cursor = $cursor ? $prefix . $cursor : null;
+
+        return sprintf('begin %s(%s%s); end;', $procedureName, $paramsString, $cursor);
+    }
+
+    /**
+     * Creates statement from procedure
+     *
+     * @param       $procedureName
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: Missing type string.

File: src/Oci8/Oci8Connection.php [hunk 5]
```diff
@@ -216,74 +216,143 @@ protected function getDoctrineDriver()
...
+    {
+        $sql = $this->createSqlFromProcedure($procedureName, $bindings, $cursorName);
+
+        return $this->getPdo()->prepare($sql);
+    }
+
+    /**
+     * Create statement from function
+     *
+     * @param       $functionName
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: Missing doc block on variable types.

**Comment by mstaack** [comment]
@yajra cleanup done

**Comment by mstaack** [comment]
@yajra will do some minor adjustment to this..please wait for emrge

**Comment by mstaack** [comment]
@yajra we should have a system independent formatter .. should i use the .php_cs config? then a lot gets changed...  or should we setup https://styleci.io ?

cheers max

**Comment by yajra** [comment]
I think `.php_cs` is not really formatting the code as expected. I think I tried several times to get the coding style of Laravel but can't get it perfectly. If you can spare some more time to setup `styleci` then I guess we can go with it.

File: src/Oci8/Oci8Connection.php [hunk 6]
```diff
@@ -196,7 +197,6 @@ public function getDoctrineConnection()
                 $data, $this->getDoctrineDriver()
             );
         }
-
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: I think the new line here is needed. No newline on return doc block but newline is needed for function returns.

**Comment by mstaack** [comment]
@yajra i fixed php cs stuff... now you just call `./bin/php-cs-fixer fix`

what do you think?

**Comment by mstaack** [comment]
@yajra can you give me more rights to this package? then i will setup styleci for us

File: README.md [hunk 7]
```diff
@@ -93,6 +94,10 @@ To use, just update `auth.php` config and set the driver to `oracle`
 - [Jimmy Felder](https://github.com/jfelder/Laravel-OracleDB)
 - [All Contributors][link-contributors]
 
+## Development
+
+- Code Style is being done via `./bin/php-cs-fixer fix`
```
Comment (yajra) [inline README.md]: I think we can remove this part since we will go with styleci?

File: composer.json [hunk 8]
```diff
@@ -19,15 +19,19 @@
     "require-dev": {
         "mockery/mockery": "~0.9.4",
         "scrutinizer/ocular": "~1.1",
-        "phpunit/phpunit": "~5.4"
+        "phpunit/phpunit": "~5.4",
+        "friendsofphp/php-cs-fixer": "^2.4"
```
Comment (yajra) [inline composer.json]: php cs fixer not needed in favor of styleci?

File: src/Oci8/Oci8Connection.php [hunk 9]
```diff
@@ -215,75 +217,140 @@ protected function getDoctrineDriver()
...
+        $results = $statement->fetchAll(PDO::FETCH_ASSOC);
+        $statement->closeCursor();
 
-            return $results;
-        }
+        return $results;
+    }
+
+    /**
+     * Creates sql command to run a procedure with bindings
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: Missing period, this will fail on styleci.

File: src/Oci8/Oci8Connection.php [hunk 10]
```diff
@@ -215,75 +217,140 @@ protected function getDoctrineDriver()
...
+        }, array_keys($bindings)));
+
+        $prefix = count($bindings) ? ',' : '';
+        $cursor = $cursor ? $prefix . $cursor : null;
+
+        return sprintf('begin %s(%s%s); end;', $procedureName, $paramsString, $cursor);
+    }
+
+    /**
+     * Creates statement from procedure
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: Missing period, this will fail on styleci.

File: src/Oci8/Oci8Connection.php [hunk 11]
```diff
@@ -215,75 +217,140 @@ protected function getDoctrineDriver()
...
+     */
+    public function createStatementFromProcedure($procedureName, array $bindings, $cursorName = false)
+    {
+        $sql = $this->createSqlFromProcedure($procedureName, $bindings, $cursorName);
+
+        return $this->getPdo()->prepare($sql);
+    }
+
+    /**
+     * Create statement from function
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: Missing period, this will fail on styleci.

File: src/Oci8/Oci8Connection.php [hunk 12]
```diff
@@ -215,75 +217,140 @@ protected function getDoctrineDriver()
...
+            $stmt->bindParam($key, $value);
         }
 
-        $query->bindParam(':result', $result, $returnType, $length);
-        $query->execute();
+        $stmt->bindParam(':result', $result, $returnType, $length);
+        $stmt->execute();
+
+        // auto convert strings like "343" and "23434.23" to int and float
+        $result = OracleTypeCaster::tryNumeric($result);
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: Do we really need to cast the result? Maybe we can just let the developer handle the casting?

File: src/Oci8/OracleTypeCaster.php [hunk 13]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace Yajra\Oci8;
+
+class OracleTypeCaster
+{
+    /**
+     * Checks for float and int values
```
Comment (yajra) [inline src/Oci8/OracleTypeCaster.php]: Missing period, this will fail on styleci.

File: src/Oci8/OracleTypeCaster.php [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+     *
+     * @param $value
+     *
```
Comment (yajra) [inline src/Oci8/OracleTypeCaster.php]: I think laravel preset does not have whitespace before `@return` doc.

**Comment by mstaack** [comment]
@yajra  fixed..

**Comment by mstaack** [comment]
@yajra styleci works great.. i just created a pr for the fix via styleci. great :D

**Comment by mstaack** [comment]
@yajra can you have a look again?

File: src/Oci8/Oci8Connection.php [hunk 15]
```diff
@@ -215,75 +217,137 @@ protected function getDoctrineDriver()
...
      *
-     * @param string $procedureName
-     * @param array $bindings
-     * @param mixed $returnType
-     * @return array
+     * @param  string $procedureName
+     * @param  array  $bindings
+     * @return bool
      */
-    public function executeProcedure($procedureName, $bindings, $returnType = PDO::PARAM_STMT)
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: Just realized that this is a breaking change. I think the changes can be targeted on 5.5 instead?

File: src/Oci8/Oci8Connection.php [hunk 16]
```diff
@@ -215,75 +217,137 @@ protected function getDoctrineDriver()
...
         $stmt->execute();
 
-        if ($returnType === PDO::PARAM_STMT) {
-            $statement = new Statement($cursor, $this->getPdo(), $this->getPdo()->getOptions());
-            $statement->execute();
-            $results = $statement->fetchAll(PDO::FETCH_ASSOC);
-            $statement->closeCursor();
+        $statement = new Statement($cursor, $this->getPdo(), $this->getPdo()->getOptions());
+        $statement->execute();
+        $results = $statement->fetchAll(PDO::FETCH_ASSOC);
```
Comment (yajra) [inline src/Oci8/Oci8Connection.php]: I think `PDO::FETCH_CLASS` is the default of Laravel now?

**Comment by yajra** [comment]
Also, maybe we should separate the PR for StyleCI changes? So many things have been updated and its getting hard to review. Thanks! :)

File: src/Oci8/Oci8Connection.php (hunk 15 above)
Comment (mstaack) [inline src/Oci8/Oci8Connection.php]: i think the old method is not being used alot b/c the implementation was way to specific as mentioned in other issues.

I think we should add this to both versions...

File: src/Oci8/Oci8Connection.php (hunk 16 above)
Comment (mstaack) [inline src/Oci8/Oci8Connection.php]: will switch

**Comment by fdalcin** [comment]
Hi.

Any prevision when this PR will be merged?

**Comment by mstaack** [comment]
@yajra how about i revert style changes, fix small issues with fetch type and we will merge this into 5.4 and 5.5?  as said above.. the current implementation ist just not used/workin....

**Comment by mstaack** [comment]
@yajra  done!

**Comment by yajra** [comment]
@mstaack thanks for the updates. 

@fdalcin are you in favor of merging this on 5.4 although it's a potential BC? Just taking other users position on this. Thanks!

Will merge this within the week if no objections raised. Thanks!

**Comment by naurojunior** [comment]
@yajra I think that it could be applied on 5.4 because this method is not working / not fully functional.

**Comment by fdalcin** [comment]
@yajra givin the current implementation I see no problem of mergin this on 5.4. As @mstaack said, currently this is not being used/working.

**Comment by yajra** [comment]
Thanks for the feedbacks. Released on v5.4.14 🚀

**Comment by mstaack** [comment]
thanks!

**Pull request merged**